### PR TITLE
fix: Apply output file prefixes to submitter GUI

### LINKED ISSUE #136: Bug: Output path is ignored and over written in submitter
### Expected Behaviour

"output path" in submitter is set as the setting render output directory setting in the render tab.

### Current Behaviour

when I submit from Blender, Output path is over written in submitter, ignored what I put in output directory in render tab.
image (37)

### Reproduction Steps

make a scene,
set render setting, put render output directory in Render tab,
open Deadline Cloud render submitter,
go to Job-specific settings,
output pass is not set as I put in the render tab.
and I need to reset output pass every time submit the job. 

### Code Snippet

*

### PR DESCRIPTION
**Pull request by miabatta**
Fixes https://github.com/aws-deadline/deadline-cloud-for-blender/issues/136.

### What was the problem/requirement? (What/Why)
Users can specify output file prefixes in the Blender UI, but they are not shown in the submitter GUI. The `OutputFileName` job parameter is hardcoded to a default value.

### What was the solution? (How)
Addeda widget to the integrated submitter for Output File Prefix. On opening the submitter, these look for a file prefix set in the Blender UI. It will display the job template's default prefix, `output_####`, if the value is unset.

### What is the impact of this change?
Users can save render output with different file names using the Blender submitter.

### How was this change tested?
Edited existing unit tests to include the new UI settings values & ran them successfully.

I manually tested the adaptor changes to make sure render output formats matched those specified in the submitter.

Changes involving the Blender UI were manually tested using different combinations of input. I'd recommend automated UI regression tests in the future to make sure no job template parameters are left out of the integrated submitter UI.

### Was this change documented?
No

### Is this a breaking change?
**Edit 03/12/24**: No

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py [hunk 1]
```diff
@@ -52,13 +52,38 @@ def create_deadline_dialog(parent=None) -> SubmitJobToDeadlineDialog:
...
     else:
-        settings.output_path = os.path.dirname(bpy.context.blend_data.filepath)
+        settings.output_directory = os.path.dirname(bpy.context.blend_data.filepath)
+
+    if bpy.path.basename(bpy.context.scene.render.filepath):
+        settings.output_file_prefix = bpy.path.basename(bpy.context.scene.render.filepath)
+
+    # When setting the output format, we can only access the extension from the Blender API.
+    # Here, we map them to formats available in the submitter GUI combobox.
+    file_formats = {
```
Comment (lucaseck) [inline src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py]: Could a mapping be created like was done previously for KeyShot where there was a [similar issue](https://github.com/aws-deadline/deadline-cloud-for-keyshot/pull/3/files)? That way the definitions of the mappings only live in one place.

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py [hunk 2]
```diff
@@ -52,13 +52,38 @@ def create_deadline_dialog(parent=None) -> SubmitJobToDeadlineDialog:
...
+        ".jp2": "JP2",
+        ".tga": "TGA",  # Can represent both TGA and RAWTGA - we leave it up to the user to specify
+        ".cin": "CINEON",
+        ".dpx": "DPX",
+        ".exr": "OPEN_EXR",  # Again, could be both OPEN_EXR and OPEN_EXR_MULTILAYER
+        ".hdr": "HDR",
+        ".tif": "TIFF",
+        ".webp": "WEBP",
+    }
+    try:
```
Comment (lucaseck) [inline src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py]: nit: Could simplify this a little with `settings.output_format = file_formats.get(bpy.context.scene.render.file_extension, "PNG")`

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/template_filling.py [hunk 3]
```diff
@@ -49,7 +49,9 @@ class BlenderSubmitterUISettings:
 
     # Paths and files.
     project_path: str = field(default="")
-    output_path: str = field(default="")
```
Comment (lucaseck) [inline src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/template_filling.py:52]: Does this need to be a breaking change by doing this rename?

**Review by lucaseck — COMMENTED** [review]
Thanks for taking this on Mia! I have one main concern where I don't think the adaptor is ever doing anything with the output format right now [here](https://github.com/aws-deadline/deadline-cloud-for-blender/blob/mainline/src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py#L175) 

It looks like the adaptor will still be [constructing the output path](https://github.com/aws-deadline/deadline-cloud-for-blender/blob/mainline/src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py#L115) with the directory/layer_camera_prefix, but never adding the format? 

The layer and camera being added into the output path automatically is probably something we want to make configurable/optional as well, but doesn't need to be included with this change in my opinion.

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py (hunk 2 above)
Comment (miabatta) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py]: Good call, will fix!

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py (hunk 1 above)
Comment (miabatta) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py]: Definitely, that'll be a quick fix

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/template_filling.py (hunk 3 above)
Comment (miabatta) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/template_filling.py:52]: For visibility: we talked about this offline and the reasoning is that this rename is a breaking change for people building off this repository.

My preference for `directory` over `path` is purely semantic so we can keep the name as-is, as long as it's clear that it's the path to the directory :)

**Comment by miabatta** [comment]
> I have one main concern where I don't think the adaptor is ever doing anything with the output format right now [here](https://github.com/aws-deadline/deadline-cloud-for-blender/blob/mainline/src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py#L175)
> 
> It looks like the adaptor will still be [constructing the output path](https://github.com/aws-deadline/deadline-cloud-for-blender/blob/mainline/src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py#L115) with the directory/layer_camera_prefix, but never adding the format?

Thank you so much for bringing this up @lucaseck - I should have noticed, but hadn't tried overriding the scene value in the submitter and verifying the final format.

It shouldn't be a huge adaptor change to include this. I'll make these changes and let you know when they're ready for review!

**Review by erico-aws — DISMISSED** [review]

File: src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py [hunk 4]
```diff
@@ -115,6 +115,11 @@ def start_render(self, data: dict) -> None:
             f"{self.output_dir}/{self.view_layer_name}_{camera}_{self.output_file_name}"
         )
         _logger.debug(f"Set output file path to {bpy.context.scene.render.filepath}")
+
+        # Add the output format
+        bpy.context.scene.render.image_settings.file_format = self.format
```
Comment (lucaseck) [inline src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py]: nit: The adaptor schema isn't an enum so I'd be curious what happens if someone isn't using the submitter but is using the adaptor and their own template that isn't a "valid" file format.

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py [hunk 5]
```diff
@@ -60,6 +60,17 @@ def create_deadline_dialog(parent=None) -> SubmitJobToDeadlineDialog:
     else:
         settings.output_path = os.path.dirname(bpy.context.blend_data.filepath)
 
+    if bpy.path.basename(bpy.context.scene.render.filepath):
+        settings.output_file_prefix = bpy.path.basename(bpy.context.scene.render.filepath)
+
+    # Not all output formats in the Blender UI are supported by the adaptor,
```
Comment (lucaseck) [inline src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py]: Why aren't they all supported? What about them makes them not supported.

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/scene_settings_widget.py [hunk 6]
```diff
@@ -32,6 +32,22 @@
 COMBO_DEFAULT_ALL_RENDERABLE_CAMERAS = "All Renderable Cameras"
 COMBO_DEFAULT_ALL_RENDERABLE_LAYERS = "All Renderable Layers"
 
+_BLENDER_OUTPUT_FORMATS: dict[str, str] = {
```
Comment (lucaseck) [inline src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/scene_settings_widget.py]: Does this need to be private?

Comment (miabatta) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/scene_settings_widget.py]: I don't think it *needs* to. I was following Keyshot which has this dictionary private but I can make it public if that might help future work.

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py (hunk 5 above)
Comment (miabatta) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py]: I might have been mistaken here; the ones that I thought weren't supported are all video encodings. I based this assumption off of the original job template's allowed values, which we now know didn't do anything...

I'll run some tests with video encodings & see if there's any reason not to include them!

File: src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py (hunk 4 above)
Comment (miabatta) [reply src/deadline/blender_adaptor/BlenderClient/render_handlers/default_blender_handler.py]: I'll need to make some changes anyway so I can test this out + add validation on the adaptor's end if necessary.

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py (hunk 5 above)
Comment (lucaseck) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py]: Ah that makes sense. Not sure how the video encodings will work right now. My guess is we would be left with a bunch of single frame videos. If they are all video only ones right now I'd say leave them out.

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/scene_settings_widget.py (hunk 6 above)
Comment (lucaseck) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/scene_settings_widget.py]: I guess what I was getting at is this is being used here: https://github.com/aws-deadline/deadline-cloud-for-blender/pull/145/files/106a3ceb1de0bd22dbd1dea0eaa489db68ee6f94#diff-afa6ff66d18ce2a86bdc1eeef3e4b44ad32dfe76e5e50a3355123cfa82346a2fR69

Would it be any cleaner to move that check within the widget file here somewhere? The format can be swapped to the default of PNG likely on first load if it's not in the dict all within this file...I think.

Comment (miabatta) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/scene_settings_widget.py]: Oh, that should be fine I believe. I'll make a separate function in `scene_settings_widget.py` for the validation

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py (hunk 5 above)
Comment (miabatta) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/open_deadline_cloud_dialog.py]: Just confirmed through testing that trying to render frames 1-3 as a video results in 3 one-frame videos. I will keep this as-is!

**Review by lucaseck — APPROVED** [review]

File: src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/scene_settings_widget.py (hunk 6 above)
Comment (lucaseck) [reply src/deadline/blender_submitter/addons/deadline_cloud_blender_submitter/scene_settings_widget.py]: Chatted offline with Mia and the output format handling will be a separate PR

**Review by erico-aws — APPROVED** [review]

**Pull request merged**